packages/amplication-server/src/core/project/project.resolver.ts

```diff
@@ -92,9 +103,75 @@ export class ProjectResolver {
   )
   @InjectContextValue(InjectableOriginParameter.UserId, "data.user.connect.id")
   async commit(@Args() args: CreateCommitArgs): Promise<Commit | null> {
+    const projectId = args.data.project.connect.id;
+
+    await this.validateProject(projectId);
+
     return this.projectService.commit(args);
   }
 
+  async validateProject(projectId: string): Promise<void> {
+    const project = await this.projectService.findUnique({
+      where: { id: projectId },
+    });
+    const workspace = await this.workspaceService.getWorkspace({
+      where: { id: project.workspaceId },
+    });
+    const subscription = await this.subscriptionService.getCurrentSubscription(
+      workspace.id
+    );
+
+    if (!subscription) {
+      const projects = await this.projectService.findProjects({
+        where: { workspace: { id: workspace.id } },
+      });
+
+      const getServicesPerProject = projects.map(async (project) => {
+        const resources = await this.resourceService.resources({
+          where: {
+            projectId: project.id,
+          },
+        });
+
+        const services = resources.filter(
+          (x) => x.resourceType === EnumResourceType.Service
+        );
+
+        return services;
+      });
+
+      const servicesPerProject = await Promise.all(getServicesPerProject);
+      const services = servicesPerProject.flat();
+
+      if (services.length > 3) {
+        throw new Error(
+          "You have reached the maximum number of services. (Upgrade your workspace plan)"
+        );
+      }
+
+      const projectServices = await this.resourceService.resources({
+        where: { projectId: project.id },
+      });
+      const promises = projectServices.map((project) =>
+        this.validateService(project)
+      );
+
+      await Promise.all(promises);
+    }
+  }
+
+  async validateService(service: Resource): Promise<void> {
+    const entities = await this.entityService.entities({
+      where: { resourceId: service.id },
+    });
+
+    if (entities.length > 7) {
+      throw new Error(
+        "You have reached the maximum number of entities. (Upgrade your workspace plan)"
+      );
+    }
+  }
```

Comment:
Could you please provide more clarity about the purpose of the validateService function? It seems the current name does not accurately describe its function in relation to pay wall validation. It would be helpful to come up with a more specific and descriptive name for this function.

Reply:
It checks that service doesn't have more that 7 entities. See the requirements: https://github.com/amplication/amplication/issues/4599

Reply:
@EugeneTseitlin I saw the requirements and i think that the name of the function `validateService` is not clear enough, its not a generic validation its the paywall validation. What do you think about rewriting the name to `paywallServiceValidation` to clarify it?

Reply:
Now I got you. Sure, name is actually too generic. Changing it.